File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/L._Подсчёт_префикс-функции/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    a: str = sys.stdin.readline().rstrip()
    n: int = len(a)
    lcp: list[int] = [0] * n
    for i in range(1, n):
        k: int = lcp[i - 1]
        while k and a[k] != a[i]: k = lcp[k - 1]
        if a[k] == a[i]: k += 1
        lcp[i] = k
    sys.stdout.write(' '.join(str(p) for p in lcp) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/L._Подсчёт_префикс-функции/solution.cpp
#include <iostream>
#include <string>
#include <vector>



void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    std::string a;
    std::cin >> a;
    int n = a.size();
    std::vector<int> lcp (n, 0);
    for (int i = 1; i < n; ++i) {
        int k = lcp[i - 1];
        while (k && a[k] != a[i]) k = lcp[k - 1];
        if (a[k] == a[i]) ++k;
        lcp[i] = k;
    }
    for (int& p : lcp) std::printf("%d ", p);
    std::printf("\n");
}


int main() {
    solution();
}